### ISSUE TITLE
Every `terraform import` command triggers reading all data sources – leads to module refactoring taking an inordinately long time

### ISSUE THREAD

**Issue by andersthorbeck** [hashicorp/terraform#32385]
### Terraform Version

```shell
Terraform v1.3.6
on darwin_amd64
+ provider registry.terraform.io/hashicorp/azuread v2.31.0
+ provider registry.terraform.io/hashicorp/azurerm v3.34.0
+ provider registry.terraform.io/integrations/github v5.11.0
```

### Use Cases

As Terraform root modules/states grow in size, they get to a point where they become too unwieldy and need to be split up until several smaller modules. In order to split up such modules, you need to run the `terraform import` command on hundreds of resources in the new module, and `terraform state rm` on the corresponding resources in the old module. Moreover, working as part of a larger team/organization, you want this migration to be completed as quickly as possible, to minimize the chances that others might be running interfering `terraform apply` commands against these two modules at the same time, which may lead to deletion of the resources you were trying to move.

Since some weeks or months ago, there seems to have been a change introduced where the `terraform import` command automatically refreshes the entire state before it actually performs the requested import. This did not use to be the case. When splitting very large Terraform states, this is very detrimental, as it triggers reads against remote state backends and providers for hundreds of resources, for _every single resource to be imported_. In other words, importing N resources has now become an O(N^2) operation. We recently attempted to split out a part of a large module into a smaller cohesive module, but aborted the attempt after the import script had taken 6 hours without an end in sight.

From my recollection, this automatic refresh as part of the `terraform import` has been observed on both the `github`, `azurerm` and `azuread` providers, so the behaviour seems to come from Terraform Core. I have not found any mention of this change in the [Terraform Core CHANGELOG](https://github.com/hashicorp/terraform/blob/main/CHANGELOG.md), so I cannot find specifically when and how the change was introduced, but I believe it was introduced at some point between at least 2022-05-03 and 2022-11-25 (probably closed to the later end).

### Attempted Solutions

I haven't really found any good workarounds to this, short of accepting that splitting up and migrating parts of terraform states will take ages, and that you're susceptible to ruining your resources in the meantime. There doesn't seem to be any option to opt out of the automatic refresh on every single call to `terraform import`.

### Proposal

I envision two possible solutions to this:

1. Introduce a flag `-no-refresh` (name up for debate) to the `terraform import` command to disable the automatic refresh before attempting the import, and leave it up to the caller to have manually performed a refresh before the call to `terraform import`.
2. Allow the `terraform import` command to take a _list_ of resources to be imported, not just one at a time, and perform the automatic refresh only once per call to the command, not before each supplied resource. The resources could be supplied via accepting multiple (`ADDR` `ID`) argument pairs, or (perhaps more sensibly) via an input file where each line contains one `ADDR` `ID` argument pair.

### References

I have not found any related GitHub issues or pull requests.

**Comment by kmoe** [comment]
Having now actually checked the code, I think the above is incorrect, and import should not be running a full refresh:
https://github.com/hashicorp/terraform/blob/main/internal/terraform/graph_builder_plan.go#L302

@andersthorbeck why do you believe a full refresh is being run? Can you share trace logs?

It's possible that another v1.3-related change is the cause of the behaviour you are observing. Could it be that reading all of the data sources takes a long time?

**Comment by gregmoy** [comment]
@kmoe its a lot of ` Reading...` for every asset then `Still reading... ` then finally `Read complete after 16s` and gets worse as you have more stuff

**Comment by kmoe** [comment]
Thanks for the example - that makes sense. Leaving this issue open as it is well described.

**Comment by Tbohunek** [comment]
Hi @kmoe, any updates on when the `terraform import -refresh=false` might become available? Thanks :)

**Comment by mishabruml** [comment]
> Hi @kmoe, any updates on when the `terraform import -refresh=false` might become available? Thanks :)

Yeah this would be very useful!

**Comment by crw** [comment]
Thanks for your interest in this issue! This is just a reminder to please avoid "+1" comments, and to use the upvote mechanism (click or add the 👍 emoji to the original post) to indicate your support for this issue. Thanks again for the feedback!

As an addendum, I am not aware of a timeline for this issue to be resolved. Thanks again for your continued interest in this issue.

**Comment by openmonk** [comment]
Hitting the same problem with a given scenario:

terraform cloud markup project creates gcp project for other terraform projects to import into empty state, but that can not be done because TF tries to read secrets during import and we get errors because those secrets are not there.

The Ugly Workaround (works only if your target is empty state):
```
cd /project/which/has/created/the/target
terraform state pull > tmp.tfstate
terraform state list -state=tmp.tfstate | grep -v -e '^google_project\.child\["child-name"\]$' | xargs -d $'\n' terraform state rm -state=tmp.tfstate
terraform state mv -state=tmp.tfstate ''^google_project\.child\["child-name"\]$' google_project.main
mv tmp.tfstate /child/project/source
cd /child/project/source
terraform init
terraform state push tmp.tfstate

```

**Comment by kieran-lowe** [comment]
Hi @kmoe, @crw - any updates on when we could see this functionality added?

**Comment by crw** [comment]
Please see the experimental feature released in 1.5, "Generated Configuration" https://developer.hashicorp.com/terraform/language/import/generating-configuration. This will allow you to import many resources at the same time and only run the plan once, which should solve the performance problem as stated in the issue description. As such, although it is not yet out of experimental the new import methodology will obviate this issue. Thanks very much for your feedback on this issue!

**Comment by andersthorbeck** [comment]
Fantastic and elegant solution! Tried it now, it works perfectly. 
@crw Thanks for meeting this use case with a better solution than any of the ones I proposed!

**Comment by thatsk** [comment]
@andersthorbeck  do we have any blog which shows with terraform 1.5 I am also seeing lagging in 1.4 but seems like this thread is saying that 1.5 solves issues so I want to try it too let me know the exact steps to follow to speed up ?

**Comment by thatsk** [comment]
I want to see a complex import block for multiple resources to be imported.

**Comment by andersthorbeck** [comment]
@thatsk The documentation was linked to in the comment which closed this issue: https://github.com/hashicorp/terraform/issues/32385#issuecomment-1603224393.

Imports can now be done in the terraform configuration itself, via the new `import` block, as opposed to having to run a CLI command for every resource to be imported.

**Comment by judithpatudith** [comment]
In case it helps we have a tutorial showing how to import a resource, which covers some edge cases and exceptions: https://developer.hashicorp.com/terraform/tutorials/state/state-import

**Comment by onetwopunch** [comment]
Can the `-refresh=false` still be a thing though? There are several resources I've seen where the import block just doesn't work. Try it on any GCP IAM resource and the ID expects a condition_title, even if one doesn't exist and won't let you provide a blank one. The ability to prevent refresh on multiple import commands is useful even with the import block solution.

**Comment by apparentlymart** [comment]
`-refresh=false` disables the refreshing of managed resources. Data resources don't get "refreshed" in the same sense: they _must_ be read for every run, because otherwise Terraform wouldn't have any data for them. Data resources are for objects managed outside of Terraform which are therefore _expected_ to change between runs.

(You may have noticed that Terraform does tend to leave behind results from reading data resources in the state after apply is complete, but those are there only to support unusual situations like debugging in `terraform console`. It isn't possible in general for Terraform to re-use data source results between runs because the object in the state might have been created by a different version of the provider and thus be unintelligble to the current version. Terraform is designed under the assumption that only managed resource data survives from one run to the next, whereas data resources are re-read fresh every time. Changing that assumption would be a much more significant undertaking than just changing the treatment of the `-refresh=false` option to also disable reading data sources.)

**Comment by kmoe** [comment]
@onetwopunch, could you open a new issue (or two) for the following?

> There are several resources I've seen where the import block just doesn't work

Which resources? It should work for all resources.

> Try it on any GCP IAM resource and the ID expects a condition_title, even if one doesn't exist and won't let you provide a blank one

Does this only happen with the `import` block, or the `terraform import` command as well? Could you provide the config you tried?

**Comment by BarnabyShearer** [comment]
Whilst I understand that data objects need to be read or they have no data; I am not clear why they need data for an unrelated resource to be successfully imported? The common case of a CLI import should just check that the destination resource dose not have state; it doesn't use anything else from the existing state, plan or data resources?

**Comment by UntiIted** [comment]
@kmoe this is not the same case as onetwopunch is referring to, but import can fail entirely because it tries to read data regardless of dependencies. Consider a scenario where a resource is created and later read as a data source (required to get IP from azurerm_public_ip).

**Comment by catalinmer** [comment]
For `terraform import` i noticed that newer version (1.4.7 in my case) has this behavior of conditioning the import operation to a successful plan. However this is not entirely the case in older versions of terraform, like v1.0.8. I was importing some vpc resources with lots of dependencies on subnets and routes and it just works in v1.0.8, while the exact same import fails in v1.4.7. Well my error was caused by "Invalid for_each argument".

I haven't tried to import all resources at once with v1.5+ because it is a complex single shot operation and I just prefer to see how the plan changes after one or few imports. And this is possible in v1.0.8

Might be a bit offtopic but if you're struggling with imports that fail when refreshing state, give it a try with an older terraform version.

**Comment by beatcracker** [comment]
As a workaround, you can comment/remove all resources except the ones you are trying to import. Worked for me in a pinch.
